Kind image loading is broken in TestDev_WithKubecontextOverride

Kind loading is broken in `TestDev_WithKubecontextOverride` - it cannot be called on its own. 

```
--- FAIL: TestDev_WithKubecontextOverride (0.91s)
/integration/TestDev_WithKubecontextOverride/skaffold_run_with_kubecontext_override
time="2020-05-15T18:02:39-07:00" level=info msg="Namespace: skaffoldbkrlk"
time="2020-05-15T18:02:39-07:00" level=info msg="[skaffold run --default-repo gcr.io/k8s-skaffold --kube-context kind-modified-context]"
    TestDev_WithKubecontextOverride/skaffold_run_with_kubecontext_override: helper.go:194: skaffold run: exit status 1, Generating tags...
         - skaffold-example -> gcr.io/k8s-skaffold/skaffold-example:v0.36.0-2130-g78b1b9fc4
        Checking cache...
         - skaffold-example: Found. Tagging
        Tags used in deployment:
         - skaffold-example -> gcr.io/k8s-skaffold/skaffold-example:fe2e30a7f7395ff45b7ece353b2ebdf137541dc7a64f3547e998cfc5fb8cf6ed
        Loading images into kind cluster nodes...
         - gcr.io/k8s-skaffold/skaffold-example:fe2e30a7f7395ff45b7ece353b2ebdf137541dc7a64f3547e998cfc5fb8cf6ed -> Failed
    TestDev_WithKubecontextOverride/skaffold_run_with_kubecontext_override: panic.go:617: loading images into kind nodes: unable to load image with kind "gcr.io/k8s-skaffold/skaffold-example:fe2e30a7f7395ff45b7ece353b2ebdf137541dc7a64f3547e998cfc5fb8cf6ed": running [kind load docker-image --name modified-context gcr.io/k8s-skaffold/skaffold-example:fe2e30a7f7395ff45b7ece353b2ebdf137541dc7a64f3547e998cfc5fb8cf6ed]
         - stdout: ""
         - stderr: "Error: unknown cluster \"modified-context\"\n"
         - cause: exit status 1, 
    --- FAIL: TestDev_WithKubecontextOverride/skaffold_run_with_kubecontext_override (0.91s)
```

I'm thoroughly puzzled still why it can pass with other tests around / before it. 
My best bet is that somehow artifact caching makes it skip the `kind load` - but I couldn't reliably reproduce it yet locally.

--- a/integration/dev_test.go
+++ b/integration/dev_test.go
@@ -266,7 +266,6 @@ func replaceInFile(target, replacement, filepath string) ([]byte, os.FileMode, e
 }
 
 func TestDev_WithKubecontextOverride(t *testing.T) {
-	t.Skip("Skipping due to error in this logic: see https://github.com/GoogleContainerTools/skaffold/issues/4198")
 	MarkIntegrationTest(t, CanRunWithoutGcp)
 
 	testutil.Run(t, "skaffold run with kubecontext override", func(t *testutil.T) {
@@ -295,7 +294,7 @@ func createModifiedKubeconfig(namespace string) ([]byte, string, error) {
 	}
 
 	contextName := "modified-context"
-	if isKind, _ := config.IsKindCluster(kubeConfig.CurrentContext); isKind {
+	if config.IsKindCluster(kubeConfig.CurrentContext) {
 		contextName = "kind-" + contextName
 	}
 

--- a/pkg/skaffold/config/util.go
+++ b/pkg/skaffold/config/util.go
@@ -199,31 +199,50 @@ func isDefaultLocal(kubeContext string) bool {
 		return true
 	}
 
-	isKind, _ := IsKindCluster(kubeContext)
-	return isKind
+	return IsKindCluster(kubeContext)
 }
 
 // IsKindCluster checks that the given `kubeContext` is talking to `kind`.
-// It also returns the name of the `kind` cluster.
-func IsKindCluster(kubeContext string) (bool, string) {
+func IsKindCluster(kubeContext string) bool {
 	switch {
 	// With kind version < 0.6.0, the k8s context
 	// is `[CLUSTER NAME]@kind`.
 	// For eg: `cluster@kind`
 	// the default name is `kind@kind`
 	case strings.HasSuffix(kubeContext, "@kind"):
-		return true, strings.TrimSuffix(kubeContext, "@kind")
+		return true
 
 	// With kind version >= 0.6.0, the k8s context
 	// is `kind-[CLUSTER NAME]`.
 	// For eg: `kind-cluster`
 	// the default name is `kind-kind`
 	case strings.HasPrefix(kubeContext, "kind-"):
-		return true, strings.TrimPrefix(kubeContext, "kind-")
+		return true
 
 	default:
-		return false, ""
+		return false
 	}
+}
+
+// KindClusterName returns the internal kind name of a kubernetes cluster.
+func KindClusterName(clusterName string) string {
+	switch {
+	// With kind version < 0.6.0, the k8s context
+	// is `[CLUSTER NAME]@kind`.
+	// For eg: `cluster@kind`
+	// the default name is `kind@kind`
+	case strings.HasSuffix(clusterName, "@kind"):
+		return strings.TrimSuffix(clusterName, "@kind")
+
+	// With kind version >= 0.6.0, the k8s context
+	// is `kind-[CLUSTER NAME]`.
+	// For eg: `kind-cluster`
+	// the default name is `kind-kind`
+	case strings.HasPrefix(clusterName, "kind-"):
+		return strings.TrimPrefix(clusterName, "kind-")
+	}
+
+	return clusterName
 }
 
 func IsUpdateCheckEnabled(configfile string) bool {

--- a/pkg/skaffold/config/util_test.go
+++ b/pkg/skaffold/config/util_test.go
@@ -291,22 +291,40 @@ func TestIsDefaultLocal(t *testing.T) {
 func TestIsKindCluster(t *testing.T) {
 	tests := []struct {
 		context        string
-		expectedName   string
 		expectedIsKind bool
 	}{
-		{context: "kind-kind", expectedName: "kind", expectedIsKind: true},
-		{context: "kind-other", expectedName: "other", expectedIsKind: true},
-		{context: "kind@kind", expectedName: "kind", expectedIsKind: true},
-		{context: "other@kind", expectedName: "other", expectedIsKind: true},
-		{context: "docker-for-desktop", expectedName: "", expectedIsKind: false},
-		{context: "not-kind", expectedName: "", expectedIsKind: false},
+		{context: "kind-kind", expectedIsKind: true},
+		{context: "kind-other", expectedIsKind: true},
+		{context: "kind@kind", expectedIsKind: true},
+		{context: "other@kind", expectedIsKind: true},
+		{context: "docker-for-desktop", expectedIsKind: false},
+		{context: "not-kind", expectedIsKind: false},
 	}
 	for _, test := range tests {
-		testutil.Run(t, "", func(t *testutil.T) {
-			isKind, name := IsKindCluster(test.context)
+		testutil.Run(t, test.context, func(t *testutil.T) {
+			isKind := IsKindCluster(test.context)
 
 			t.CheckDeepEqual(test.expectedIsKind, isKind)
-			t.CheckDeepEqual(test.expectedName, name)
+		})
+	}
+}
+
+func TestKindClusterName(t *testing.T) {
+	tests := []struct {
+		kubeCluster  string
+		expectedName string
+	}{
+		{kubeCluster: "kind", expectedName: "kind"},
+		{kubeCluster: "kind-kind", expectedName: "kind"},
+		{kubeCluster: "kind-other", expectedName: "other"},
+		{kubeCluster: "kind@kind", expectedName: "kind"},
+		{kubeCluster: "other@kind", expectedName: "other"},
+	}
+	for _, test := range tests {
+		testutil.Run(t, test.kubeCluster, func(t *testutil.T) {
+			kindCluster := KindClusterName(test.kubeCluster)
+
+			t.CheckDeepEqual(test.expectedName, kindCluster)
 		})
 	}
 }

--- a/pkg/skaffold/runner/deploy.go
+++ b/pkg/skaffold/runner/deploy.go
@@ -28,6 +28,7 @@ import (
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/config"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+	kubectx "github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/context"
 )
 
 func (r *SkaffoldRunner) Deploy(ctx context.Context, out io.Writer, artifacts []build.Artifact) error {
@@ -55,7 +56,19 @@ See https://skaffold.dev/docs/pipeline-stages/taggers/#how-tagging-works`)
 		return fmt.Errorf("unable to connect to Kubernetes: %w", err)
 	}
 
-	if isKind, kindCluster := config.IsKindCluster(r.runCtx.KubeContext); isKind {
+	if config.IsKindCluster(r.runCtx.KubeContext) {
+		currentCfg, err := kubectx.CurrentConfig()
+		if err != nil {
+			return fmt.Errorf("unable to get kubernetes config: %w", err)
+		}
+
+		currentContext, present := currentCfg.Contexts[r.runCtx.KubeContext]
+		if !present {
+			return fmt.Errorf("unable to get current kubernetes context: %w", err)
+		}
+
+		kindCluster := config.KindClusterName(currentContext.Cluster)
+
 		// With `kind`, docker images have to be loaded with the `kind` CLI.
 		if err := r.loadImagesInKindNodes(ctx, out, kindCluster, artifacts); err != nil {
 			return fmt.Errorf("loading images into kind nodes: %w", err)

